File: Models/PhotoItem.cs
using System.ComponentModel;
using System.Runtime.CompilerServices;
using System.Windows.Media.Imaging;

namespace iPhotoImporter.Models;

/// <summary>
/// Representa una foto o vídeo, tant d'un dispositiu MTP com d'una carpeta local.
/// </summary>
public class PhotoItem : INotifyPropertyChanged
{
    private bool _isSelected;
    private BitmapSource? _thumbnail;
    private bool _isHighlighted;
    private string? _duplicateGroupId;
    private string? _location;

    // === Extensions centralitzades (matching macOS) ===

    /// <summary>Extensions d'imatge estàndard</summary>
    public static readonly HashSet<string> ImageExtensions = new(StringComparer.OrdinalIgnoreCase)
    {
        ".jpg", ".jpeg", ".png", ".bmp", ".gif", ".webp", ".tiff", ".tif", ".heic", ".heif"
    };

    /// <summary>Extensions RAW de càmera</summary>
    public static readonly HashSet<string> RawExtensions = new(StringComparer.OrdinalIgnoreCase)
    {
        ".cr2", ".cr3", ".nef", ".arw", ".dng", ".raf", ".orf", ".rw2", ".pef", ".srw", ".rwl"
    };

    /// <summary>Extensions de formats moderns</summary>
    public static readonly HashSet<string> ModernExtensions = new(StringComparer.OrdinalIgnoreCase)
    {
        ".avif", ".jxl", ".psd"
    };

    /// <summary>Extensions de vídeo</summary>
    public static readonly HashSet<string> VideoExtensions = new(StringComparer.OrdinalIgnoreCase)
    {
        ".mp4", ".mov", ".avi", ".mkv", ".m4v", ".webm", ".3gp", ".mts", ".m2ts"
    };

    /// <summary>Totes les extensions suportades</summary>
    public static readonly HashSet<string> AllExtensions = new(
        ImageExtensions
            .Concat(RawExtensions)
            .Concat(ModernExtensions)
            .Concat(VideoExtensions),
        StringComparer.OrdinalIgnoreCase);

    // === Propietats bàsiques ===

    /// <summary>Ruta completa (local o MTP)</summary>
    public required string FullPath { get; init; }

    /// <summary>Nom del fitxer (ex: IMG_20240101.jpg)</summary>
    public required string FileName { get; init; }

    /// <summary>Data de captura de la foto</summary>
    public DateTime? DateTaken { get; set; }

    /// <summary>Mida del fitxer en bytes</summary>
    public long SizeBytes { get; init; }

    /// <summary>Indica si és un fitxer local (true) o MTP (false)</summary>
    public bool IsLocal { get; init; }

    /// <summary>Amplada de la imatge original (si disponible)</summary>
    public int PixelWidth { get; set; }

    /// <summary>Alçada de la imatge original (si disponible)</summary>
    public int PixelHeight { get; set; }

    /// <summary>Rotació del vídeo en graus (0, 90, 180, 270)</summary>
    public int VideoRotation { get; set; }

    // === Propietats de deduplicació ===

    /// <summary>Hash MD5 del fitxer complet</summary>
    public string? Md5Hash { get; set; }

    /// <summary>Hash perceptual (8x8 grayscale average hash)</summary>
    public ulong? PerceptualHash { get; set; }

    /// <summary>Fingerprint EXIF (SHA256 de datetime+camera+dimensions)</summary>
    public string? ExifFingerprint { get; set; }

    /// <summary>ID del grup de duplicats (prefix: md5-, phash-, exif-)</summary>
    public string? DuplicateGroupId
    {
        get => _duplicateGroupId;
        set => SetField(ref _duplicateGroupId, value);
    }

    // === Propietats GPS/Localització ===

    /// <summary>Latitud GPS des d'EXIF</summary>
    public double? GpsLatitude { get; set; }

    /// <summary>Longitud GPS des d'EXIF</summary>
    public double? GpsLongitude { get; set; }

    /// <summary>Localització geocodificada (ex: "Barcelona, Catalunya")</summary>
    public string? Location
    {
        get => _location;
        set => SetField(ref _location, value);
    }

    // === Propietats UI (observables) ===

    /// <summary>Seleccionat per l'usuari a la UI</summary>
    public bool IsSelected
    {
        get => _isSelected;
        set => SetField(ref _isSelected, value);
    }

    /// <summary>Ressaltat visualment (ex: element actiu al visor)</summary>
    public bool IsHighlighted
    {
        get => _isHighlighted;
        set => SetField(ref _isHighlighted, value);
    }

    /// <summary>Miniatura de la foto (pot ser null si no disponible)</summary>
    public BitmapSource? Thumbnail
    {
        get => _thumbnail;
        set => SetField(ref _thumbnail, value);
    }

    // === Propietats computades ===

    /// <summary>Indica si és un vídeo</summary>
    public bool IsVideo
    {
        get
        {
            var ext = System.IO.Path.GetExtension(FileName);
            return VideoExtensions.Contains(ext);
        }
    }

    /// <summary>Indica si és un format RAW</summary>
    public bool IsRaw
    {
        get
        {
            var ext = System.IO.Path.GetExtension(FileName);
            return RawExtensions.Contains(ext);
        }
    }

    /// <summary>Indica si és una imatge (no vídeo)</summary>
    public bool IsImage => !IsVideo;

    // === INotifyPropertyChanged ===

    public event PropertyChangedEventHandler? PropertyChanged;

    protected void OnPropertyChanged([CallerMemberName] string? propertyName = null)
        => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));

    private bool SetField<T>(ref T field, T value, [CallerMemberName] string? propertyName = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value)) return false;
        field = value;
        OnPropertyChanged(propertyName);
        return true;
    }
}

File: MacOS/iPhotoManager/Models/PhotoItem.swift
import Foundation
import AppKit
import AVFoundation
import ImageCaptureCore

/// Represents a photo or video item, from a local folder or an MTP device.
@Observable
final class PhotoItem: Identifiable, Hashable {

    // MARK: - Identity

    let id: String
    let fullPath: String
    let fileName: String
    let isLocal: Bool

    // MARK: - Metadata

    var dateTaken: Date?
    let sizeBytes: Int64
    var pixelWidth: Int = 0
    var pixelHeight: Int = 0

    /// Video rotation in degrees (0, 90, 180, 270) from track header metadata.
    var videoRotation: Int = 0

    /// Reverse-geocoded location name (city).
    var location: String?

    /// Raw GPS coordinates (for deferred geocoding).
    var gpsLatitude: Double?
    var gpsLongitude: Double?

    /// Reference to the device camera file (only for device items, nil for local).
    var cameraFile: ICCameraFile?

    // MARK: - Deduplication

    /// MD5 hash of file content (exact duplicate detection).
    var md5Hash: String?

    /// Perceptual hash — 64-bit average hash of visual content.
    var perceptualHash: UInt64?

    /// EXIF fingerprint — hash of camera + datetime + dimensions.
    var exifFingerprint: String?

    /// Group ID for duplicates (shared by all items in same duplicate group).
    var duplicateGroupId: String?

    // MARK: - UI State

    var isSelected: Bool = false
    var isHighlighted: Bool = false
    var thumbnail: NSImage?

    // MARK: - Computed

    var isVideo: Bool {
        let ext = (fileName as NSString).pathExtension.lowercased()
        return Self.videoExtensions.contains(ext)
    }

    var isImage: Bool {
        let ext = (fileName as NSString).pathExtension.lowercased()
        return Self.imageExtensions.contains(ext)
    }

    var isRaw: Bool {
        let ext = (fileName as NSString).pathExtension.lowercased()
        return Self.rawExtensions.contains(ext)
    }

    var fileSizeFormatted: String {
        ByteCountFormatter.string(fromByteCount: sizeBytes, countStyle: .file)
    }

    // MARK: - Supported extensions

    static let imageExtensions: Set<String> = [
        // Standard
        "jpg", "jpeg", "png", "bmp", "gif", "webp", "tiff", "tif",
        // Apple
        "heic", "heif",
        // Modern
        "avif", "jxl",
        // RAW — Canon
        "cr2", "cr3", "crw",
        // RAW — Nikon
        "nef", "nrw",
        // RAW — Sony
        "arw", "srf", "sr2",
        // RAW — Adobe
        "dng",
        // RAW — Fujifilm
        "raf",
        // RAW — Olympus / OM System
        "orf",
        // RAW — Panasonic
        "rw2",
        // RAW — Pentax
        "pef",
        // RAW — Samsung
        "srw",
        // RAW — Leica
        "rwl",
        // Adobe
        "psd",
    ]

    static let videoExtensions: Set<String> = [
        "mp4", "mov", "avi", "mkv",
        "m4v", "webm", "3gp",
        "mts", "m2ts", "ts",
    ]

    /// Extensions que són RAW (per optimitzar amb preview embegut)
    static let rawExtensions: Set<String> = [
        "cr2", "cr3", "crw", "nef", "nrw", "arw", "srf", "sr2",
        "dng", "raf", "orf", "rw2", "pef", "srw", "rwl",
    ]

    static let allExtensions: Set<String> = imageExtensions.union(videoExtensions)

    // MARK: - Init

    init(fullPath: String, fileName: String, dateTaken: Date? = nil,
         sizeBytes: Int64, isLocal: Bool = true) {
        self.id = fullPath
        self.fullPath = fullPath
        self.fileName = fileName
        self.dateTaken = dateTaken
        self.sizeBytes = sizeBytes
        self.isLocal = isLocal
    }

    /// Convenience init for device items.
    init(cameraFile: ICCameraFile, deviceId: String) {
        let name = cameraFile.name ?? "unknown"
        let folder = cameraFile.parentFolder?.name ?? ""
        // Unicitat de l'id: a l'iPhone (MTP) hi pot haver fitxers DIFERENTS amb el
        // MATEIX nom — el comptador IMG_#### es reinicia (després de IMG_9999),
        // hi ha fotos importades d'altres fonts, o l'estructura es presenta plana
        // (mateix `parentFolder.name`). Si l'id col·lisiona, com que `==`/`hash`
        // es basen en `id`, dues fotos passen per "la mateixa": `firstIndex(of:)`
        // retorna l'altra i SwiftUI (Identifiable per id) reusa la cel·la → al
        // visor surt una i la canvia per l'altra. Afegim mida + data de creació
        // (continguts diferents → mides/dates diferents) perquè l'id sigui únic.
        let size = cameraFile.fileSize
        let ts = Int((cameraFile.creationDate ?? .distantPast).timeIntervalSince1970)
        let path = "device://\(deviceId)/\(folder)/\(name)#\(size)-\(ts)"
        self.id = path
        self.fullPath = path
        self.fileName = name
        self.dateTaken = cameraFile.creationDate
        self.sizeBytes = size
        self.isLocal = false
        self.cameraFile = cameraFile
    }

    // MARK: - Hashable

    static func == (lhs: PhotoItem, rhs: PhotoItem) -> Bool {
        lhs.id == rhs.id
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}
